drgn/helpers/linux/dmesg.py

```diff
@@ -0,0 +1,158 @@
+    text_data_ring = prb.text_data_ring
+    text_data_ring_size = 1 << text_data_ring.size_bits
+
+    head_id = desc_ring.head_id.counter
+    current_id = desc_ring.tail_id.counter
+
+    desc_committed = prog["desc_committed"]
+    desc_finalized = prog["desc_finalized"]
+
+    result = []
+
+    def record_committed(current_id: Object, state_var: Object) -> bool:
+        state_desc_id = state_var & DESC_ID_MASK
+        state = 3 & (state_var >> DESC_FLAGS_SHIFT)
+        return (current_id == state_desc_id) and (
+            state == desc_committed or state == desc_finalized
+        )
+
+    def dump_record(current_id: Object) -> None:
+        idx = current_id % desc_ring_count
+        desc = desc_ring.descs[idx]
+        info = desc_ring.infos[idx]
+        text_len = info.text_len
+        if not record_committed(current_id, desc.state_var.counter):
+            return
+
+        lpos_begin = desc.text_blk_lpos.begin % text_data_ring_size
+        lpos_next = desc.text_blk_lpos.next % text_data_ring_size
+        lpos_begin += ulong_size
+
+        if lpos_begin == lpos_next:
+            return
+        if lpos_begin > lpos_next:
+            lpos_begin -= lpos_begin
+        if lpos_next - lpos_begin < text_len:
+            text_len = lpos_next - lpos_begin
+
+        text = prog.read(text_data_ring.data + lpos_begin, text_len)
+
+        context = {}
+        subsystem = info.dev_info.subsystem.string_()
+        device = info.dev_info.device.string_()
+        if subsystem:
+            context[b"SUBSYSTEM"] = subsystem
+        if device:
+            context[b"DEVICE"] = device
+
+        result.append(
+            PrintkRecord(
+                text=text,
+                facility=info.facility.value_(),
+                level=info.level.value_(),
+                seq=info.seq.value_(),
+                timestamp=info.ts_nsec.value_(),
+                continuation=bool(info.flags.value_() & LOG_CONT),
+                context=context,
+            )
+        )
+
+    while current_id != head_id:
+        dump_record(current_id)
+        current_id = (current_id + 1) & DESC_ID_MASK
+
+    dump_record(current_id)
+    return result
+
+
+def dump_dmesg_structured(prog: Program) -> List[PrintkRecord]:
+    result = []
+    printk_logp_type = prog.type("struct printk_log *")
+    current_idx = prog["log_first_idx"]
+    next_idx = prog["log_next_idx"]
+    seq = prog["log_first_seq"]
+    LOG_CONT = prog["LOG_CONT"].value_()
+
+    while current_idx != next_idx:
+        log = cast(printk_logp_type, prog["log_buf"] + current_idx)
+        text_dict = prog.read(log + 1, log.text_len + log.dict_len)
+        text_len = int(log.text_len)
+        dict_len = int(log.dict_len)
+        text = text_dict[:text_len]
+        context: Dict[bytes, bytes] = {}
+        if dict_len:
+            context = dict(
+                item.split(b"=", 1) for item in text_dict[text_len:].split(b"\0")
```

Comment:
I guess mypy isn't smart enough to accept this code, so we can probably go back to the way you originally did this.